src/dbnode/persist/fs/registry.go

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/clock"
+	"github.com/m3db/m3/src/x/close"
+	"github.com/m3db/m3/src/x/instrument"
+	"github.com/m3db/m3/src/x/watch"
+	"go.uber.org/zap"
+)
+
+const (
+	notifyAndRemoveTimeout       = time.Minute
+	notifyAndRemoveCheckInterval = time.Second
+)
+
+type fileRegistry struct {
+	sync.RWMutex
+
+	entries map[string]*fileStatus
+
+	clockOpts clock.Options
+	logger    *zap.Logger
+}
+
+// NewFileRegistry creates a new file registry.
+func NewFileRegistry(
+	clockOpts clock.Options,
+	instrumentOpts instrument.Options,
+) FileRegistry {
+	return &fileRegistry{
+		entries:   make(map[string]*fileStatus),
+		clockOpts: clockOpts,
+		logger:    instrumentOpts.Logger(),
+	}
+}
+
+func (r *fileRegistry) Open(filePath string) (File, error) {
+	entry, err := r.newFileStatusOrExisting(filePath)
+	if err != nil {
+		return nil, err
+	}
+	return entry.Open()
+}
+
+func (r *fileRegistry) OpenFile(filePath string, flag int, perm os.FileMode) (File, error) {
+	entry, err := r.newFileStatusOrExisting(filePath)
+	if err != nil {
+		return nil, err
+	}
+	return entry.OpenFile(flag, perm)
+}
+
+func (r *fileRegistry) Remove(filePath string) error {
+	pathAbs, err := filepath.Abs(filePath)
+	if err != nil {
+		return err
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	status, ok := r.entries[pathAbs]
+	if !ok {
+		// Just fine, no files open.
+		return os.Remove(pathAbs)
+	}
+
+	// Delete from the registry.
+	delete(r.entries, pathAbs)
+
+	// Notify open file holders that this file is being removed
+	// and perform a delayed remove which will delete the file
+	// after all references are released.
+	go func() {
```

Comment:
couple of questions about this:
(1) any situation where callers would want the ability to wait the file was removed before `Remove()` returns?
(2) along the same lines, you're removing the entry from the map above, but the file might not be remove because of an error returned in the goroutine. may cause funky state?

Reply:
Yeah this feels racey and off to me.

Scenario:

```
1. Cleanup manager calls Remove() on some file. File exists in map, gets deleted from map, spawns background routine, then returns nil.
2. Background routine fails (for whatever reason).
3. Cleanup manager comes around again, finds the files on disk (because they haven't been deleted), then calls delete again, its not in this map, so we just issue a `os.Remove()` even though the filemanager may still have a handle on it
```

Similar scenario applies for the merge/compaction logic I think.

On top of that, if we have this operation async we can no longer serialize access to our data and snapshot files. I.E right now the logic is:

1. Data flush
2. Data cold flush
3. Data snapshot
4. Cleanup

perform in a synchronous loop with no overlap. With this background goroutine we lose all control over that and the processes will be competing.

An alternative approach that might work nicely is to just provide a batch API, so something like

```golang
func (r *fileRegistry) Remove(filePaths []string) error {
    for _, fPath := range filePaths {
        // If not in map, remove and continue
        // Call NotifyAndRemove() and wait for success (can have some parallelism here)
        // If success, remove from map, other keep it in and pass back an []error or something for all the ones that failed
    }

    wg.Wait() // Wait for all of them to complete or error out
}
```

So now the map state is always "correct" hopefully, we don't end up with overlapping processes, and it should work nicely with the compaction logic too because that can just accumulate a list of files to be removed and then delete them all at once at the end

This may not be the correct answer but I think the existing one may cause more harm than good by introducing an async layer here